### PR TITLE
[MERGE][REF] product, sale: tests

### PR DESCRIPTION
**Pull request by Feyensv**
* Clean tests setups
  * Use setUpClass instead of setUp
  * remove useless test data or move it to the place(s) of use
    * when possible, use existing test data instead of creating another custom setup for the current test 
  * remove unnecessary dependencies on accounting common
  * make tests work without demo data (when not relying on accounting common)
  * run base product/sale tests without triggering mail logic
  * create records in batch to improve performance (reduced queries, batched operations, ...)
* Create the base of a clean test architecture for those modules (with dedicated commons in the right modules)
* Improve some tests (mainly prices/pricelist tests in sale)
* Add `mute_logger` contexts in some tests to reduce useless test logs
* Try to improve tests consistency (fixed test currency, archiving existing demo data to avoid side-effects, ...)
* Move most `product` tests `post_install` to benefit from the split of `post_install` tests on runbot.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/sale/tests/test_sale_prices.py [hunk 1]
```diff
@@ -0,0 +1,961 @@
...
+            'order_id': self.empty_order.id,
+            'product_id': self.product.id,
+        })
+
+        self.assertEqual(order_line.pricelist_item_id, self.pricelist.item_ids)
+        self.assertEqual(order_line.price_subtotal, 81, "Second pricelist rule not applied")
+        self.assertEqual(order_line.discount, 19, "Second discount not applied")
+
+    def test_pricelist_with_another_currency(self):
+        """ Test prices are correctly applied with a pricelist with an other currency"""
```
Comment (Demesmaeker) [inline addons/sale/tests/test_sale_prices.py:297]:
```suggestion
        """ Test prices are correctly applied with a pricelist with another currency"""
```

**Review by Demesmaeker — COMMENTED** [review]
:duck:

File: odoo/addons/base/tests/common.py [hunk 2]
```diff
@@ -7,28 +7,99 @@
...
+        super().setUpClass()
+
+        # Enforce the use of USD as main currency
+        # unless modified in inherited class(es)
+        cls._use_currency('USD')
+
+        # By default, we consider mail logic
+        # should be tested by mail tests, and not by
+        # other modules tests
+        # Hack to use with_context and avoir manual context dict modification here
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply odoo/addons/base/tests/common.py:32]: :grin:

File: addons/product/tests/common.py [hunk 3]
```diff
@@ -1,118 +1,181 @@
...
+    BaseCommon,  # enforce constant test currency (USD)
+    UomCommon,
+):
 
     @classmethod
     def setUpClass(cls):
-        super(TestProductCommon, cls).setUpClass()
+        super().setUpClass()
+
+        cls.env = cls.env['base'].with_context(**DISABLED_MAIL_CONTEXT).env
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/product/tests/common.py:21]: Indeed, leftover from the time where there was a BaseCommon and a BaseCurrencyCommon :eyes:

File: addons/product/tests/test_pricelist.py [hunk 4]
```diff
@@ -1,51 +1,71 @@
...
+    #   test percentage (discount)
+    #   test formula
+    #       surcharge
+    #       rounding
+    #       min&max margins
+    #       different bases
+    #           sales price,
+    #           cost, (same or different cost currency)
+    #           pricelist (same & different currency)
+
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/product/tests/test_pricelist.py:58]: To c/p in your future pricelist test task :D

[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/product/tests/test_pricelist.py:58]: hm... Not in a task specification, so I didn't read that =P

File: addons/uom/tests/common.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
+# -*- coding: utf-8 -*-
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/uom/tests/common.py:1]: I don't care and I prefer consistency, even if it's useless :D

**Review by Julien00859 — COMMENTED** [review]
While I'm usually all in for generic solutions, I have a mixed feeling about those two new `BaseCommon` and `BaseUsersCommon` classes in `common.py`.

To me the two classes just do too little: just enforcing USD, disabling some automatic stuff from mail, create a few users/partners. In the same time, they do too much, e.g. not all tests need to enforce the USD currency but you pay the query anyway (but that's fine as it is in a setUpClass).

Introducing two new classes gives too many options to developers, it was confusing at the time we had to choose between `TransactionCase` and `SavepointCase` and now I can imagine people asking themselves:
> should I use a TransactionCase, a BaseCommon or a BaseUsersCommon?
> should I refactor my tests to use BaseCommon instead of TransactionCase?

I know there are *many* tests that use a `setUp` created partner to play with. If those tests were to use `BaseCommon` instead, we would have a single attribute to remember (`self.partner`) and it would be faster because created in a `setUpClass` instead of inside a `setUp`.

On the other hand, if we continue *not to have* a BaseCommon-like class, people would just continue to duplicate the same setUp a partner and a few users code...

So yes... mixed feelings...

Otherwise, implementation wise I'm ok :)

**Comment by Feyensv** [comment]
TBH, I don't see this as the final version but as a step towards a simplified tests architecture.

I didn't want to go too far directly, but in the future, this BaseCommon should ideally setup a whole new company to avoid relying on demo data at all (and maybe some other things, I've kept the BaseUsersCommon separate for now because for my current use, most tests do not need those test users, but it might be merged in the BaseCommon someday)

**Comment by Julien00859** [comment]
Well... I'm not sure that setting up an entire company at each setUpClass is a good idea...

**Comment by Feyensv** [comment]
That's what the accounting common does :D
(and if we really want to consider tests independent of demo data, it would be necessary ^^)

**Comment by Julien00859** [comment]
The thing I don't like is to re-create the stuff on each setUpClass... if only we could had a transaction for creating that company and then *sub*-transactions to run each tests. We would setup the company once, using as much queries as we want, each test would run in a transaction *within* that company transaction and that sub-transaction would be rollbacked at the end of the test (restoring the original company state). At the end of the test-suite, the first transaction is rollbacked too, leaving no trace in the database.

But sub-transactions is not a thing.

**Comment by Feyensv** [comment]
That's why i proposed (again) the use of "tests" data specified by module in the manifest (and allowing us to drop the use of demo data in tests), but didn't convince much people atm :D.

**Comment by Feyensv** [comment]
@robodoo r+ rebase-merge

**Comment by xmo-odoo** [comment]
@robodoo retry

Runbot apparently kinda just died over the weekend, so timeout is likely not legit.

**Pull request closed without merging**